From 3cd1c17402800b5c1cd83a4f6616037bf4c09a08 Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@nbd.name>
Date: Mon, 19 Aug 2024 20:21:04 +0200
Subject: [PATCH 5/5] wifi: mt76: mt7915: hold dev->mt76.mutex while disabling
 tx worker

Prevent racing against other functions disabling the same worker

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---
 mt7915/mac.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -1538,12 +1538,14 @@ void mt7915_mac_reset_work(struct work_s
 		set_bit(MT76_RESET, &phy2->mt76->state);
 		cancel_delayed_work_sync(&phy2->mt76->mac_work);
 	}
+
+	mutex_lock(&dev->mt76.mutex);
+
 	mt76_worker_disable(&dev->mt76.tx_worker);
 	mt76_for_each_q_rx(&dev->mt76, i)
 		napi_disable(&dev->mt76.napi[i]);
 	napi_disable(&dev->mt76.tx_napi);
 
-	mutex_lock(&dev->mt76.mutex);
 
 	if (mtk_wed_device_active(&dev->mt76.mmio.wed))
 		mtk_wed_device_stop(&dev->mt76.mmio.wed);
